go/vt/vtgate/engine/pullout_subquery.go

```diff
@@ -0,0 +1,172 @@
+
+import (
+	"fmt"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/vterrors"
+
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+)
+
+var _ Primitive = (*PulloutSubquery)(nil)
+
+// PulloutSubquery executes a "pulled out" subquery and stores
+// the results in a bind variable.
+type PulloutSubquery struct {
+	Opcode         PulloutOpcode
+	SubqueryResult string
+	HasValues      string
+	IsEmpty        string
+	Subquery       Primitive
+	Underlying     Primitive
+}
+
+// RouteType returns a description of the query routing type used by the primitive
+func (ps *PulloutSubquery) RouteType() string {
+	return ps.Opcode.String()
+}
+
+// Execute satisfies the Primtive interface.
+func (ps *PulloutSubquery) Execute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	combinedVars, err := ps.execSubquery(vcursor, bindVars)
+	if err != nil {
+		return nil, err
+	}
+	return ps.Underlying.Execute(vcursor, combinedVars, wantfields)
+}
+
+// StreamExecute performs a streaming exec.
+func (ps *PulloutSubquery) StreamExecute(vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool, callback func(*sqltypes.Result) error) error {
+	combinedVars, err := ps.execSubquery(vcursor, bindVars)
+	if err != nil {
+		return err
+	}
+	return ps.Underlying.StreamExecute(vcursor, combinedVars, wantfields, callback)
+}
+
+// GetFields fetches the field info.
+func (ps *PulloutSubquery) GetFields(vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	combinedVars := make(map[string]*querypb.BindVariable, len(bindVars)+1)
+	for k, v := range bindVars {
+		combinedVars[k] = v
+	}
+	switch ps.Opcode {
+	case PulloutValue:
+		combinedVars[ps.SubqueryResult] = sqltypes.NullBindVariable
+	case PulloutIn, PulloutNotIn:
+		combinedVars[ps.HasValues] = sqltypes.Int64BindVariable(0)
+		combinedVars[ps.IsEmpty] = sqltypes.Int64BindVariable(1)
+		combinedVars[ps.SubqueryResult] = &querypb.BindVariable{
+			Type:   querypb.Type_TUPLE,
+			Values: []*querypb.Value{sqltypes.ValueToProto(sqltypes.NewInt64(0))},
+		}
+	case PulloutExists:
+		combinedVars[ps.SubqueryResult] = sqltypes.Int64BindVariable(0)
+	}
+	return ps.Underlying.GetFields(vcursor, combinedVars)
+}
+
+func (ps *PulloutSubquery) execSubquery(vcursor VCursor, bindVars map[string]*querypb.BindVariable) (map[string]*querypb.BindVariable, error) {
+	result, err := ps.Subquery.Execute(vcursor, bindVars, false)
+	if err != nil {
+		return nil, err
+	}
+	combinedVars := make(map[string]*querypb.BindVariable, len(bindVars)+1)
+	for k, v := range bindVars {
+		combinedVars[k] = v
+	}
+	switch ps.Opcode {
+	case PulloutValue:
+		switch len(result.Rows) {
+		case 0:
+			combinedVars[ps.SubqueryResult] = sqltypes.NullBindVariable
+		case 1:
+			if len(result.Rows[0]) != 1 {
+				return nil, vterrors.New(vtrpcpb.Code_INVALID_ARGUMENT, "subquery returned more than one column")
```

Comment:
Here you are checking the number of rows, but the error talks about the number of columns. By design?

Reply:
actually, result.Rows[0] gets us the first row, which is a slice(array) consisting of column values.